src/core/struct_tree.js

```diff
@@ -13,29 +13,49 @@
  * limitations under the License.
  */
 
-import { Dict, isName, Name, Ref } from "./primitives.js";
+import { Dict, isName, Name, Ref, RefSetCache } from "./primitives.js";
 import { stringToPDFString, warn } from "../shared/util.js";
 import { NumberTree } from "./name_number_tree.js";
 
 const MAX_DEPTH = 40;
 
 const StructElementType = {
-  PAGE_CONTENT: "PAGE_CONTENT",
-  STREAM_CONTENT: "STREAM_CONTENT",
-  OBJECT: "OBJECT",
-  ELEMENT: "ELEMENT",
+  PAGE_CONTENT: 0,
+  STREAM_CONTENT: 1,
+  OBJECT: 2,
+  ANNOTATION: 3,
+  XOBJECT: 4,
+  ELEMENT: 5,
 };
 
 class StructTreeRoot {
   constructor(rootDict) {
     this.dict = rootDict;
     this.roleMap = new Map();
+    this.structParentIds = null;
   }
 
   init() {
     this.readRoleMap();
   }
 
+  #addIdToPage(pageRef, id, type) {
+    if (id < 0) {
+      return;
+    }
```

Comment:
The `pageRef` can be `null` in some rare cases, so you definitely want to return early when `!(pageRef instanceof Ref)` as well here; see https://github.com/mozilla/pdf.js/blob/92f7653cfbc2b8e824991f496564cf17fc3ee09b/src/core/catalog.js#L1151-L1161